hopp/tools/design/wind/turbine_library_interface_tools.py

```diff
@@ -0,0 +1,242 @@
+
+def extract_power_curve(turbine_specs: dict, model_name: str):
+    """Creates power-curve for turbine based on available data and formats it for the corresponding simulation model.
+
+    Args:
+        turbine_specs (dict): turbine specs loaded from turbine-models library.
+        model_name (str): wind simulation model, either "pysam" or "floris".
+
+    Raises:
+        ValueError: if turbine data doesn't have the minimum required power-curve information.
+        ValueError: if model name is not either 'pysam' or 'floris'
+
+    Returns:
+        dict: power-curve dictionary formatted for the corresponding ``model_name``.
+    """
+
+    if model_name not in ("floris", "pysam"):
+        raise ValueError(f"model_name {model_name} is invalid, options are either 'floris' or 'pysam'.")
+    turbine_specs["power_curve"] = turbine_specs["power_curve"].dropna()
+    wind_speeds = np.nan_to_num(turbine_specs["power_curve"]["wind_speed_ms"].to_list())
+    turbine_curve_cols = turbine_specs["power_curve"].columns.to_list()
+    
+    has_cp_curve = "cp" in turbine_curve_cols
+    has_power_curve = "power_kw" in turbine_curve_cols
+    has_ct_curve = "ct" in turbine_curve_cols
+    
+    if not has_cp_curve and not has_power_curve:
+        turbine_name = turbine_specs["name"]
+        raise ValueError(f"turbine {turbine_name} does not have minimum required power curve data (needs either power_kw or cp)")
+    
+    if has_cp_curve:
+        cp_curve = np.array(turbine_specs["power_curve"]["cp"].to_list())
+        cp_curve = np.nan_to_num(cp_curve)
+        cp_curve = np.where(cp_curve<0,0,cp_curve).tolist()
+
+    if has_power_curve and has_cp_curve:
+        power_curve_kw = np.array(turbine_specs["power_curve"]["power_kw"].to_list())
+        power_curve_kw = np.nan_to_num(power_curve_kw)
+        power_curve_kw = np.where(power_curve_kw<0,0,power_curve_kw)
+        power_curve_kw = np.where(power_curve_kw>turbine_specs["rated_power"],turbine_specs["rated_power"],power_curve_kw).tolist()
+        
+        cp_curve = np.array(turbine_specs["power_curve"]["cp"].to_list())
+        cp_curve = np.nan_to_num(cp_curve)
+        cp_curve = np.where(cp_curve<0,0,cp_curve).tolist()
+        
+    if has_cp_curve and not has_power_curve:
+        cp_curve = np.array(turbine_specs["power_curve"]["cp"].to_list())
+        cp_curve = np.nan_to_num(cp_curve)
+        cp_curve = np.where(cp_curve<0,0,cp_curve).tolist()
+        power_curve_kw = curve_tools.calculate_power_from_cp(wind_speeds,cp_curve,turbine_specs["rotor_diameter"],turbine_specs["rated_power"])
+    
+    if has_power_curve and not has_cp_curve:
+        power_curve_kw = np.array(turbine_specs["power_curve"]["power_kw"].to_list())
+        power_curve_kw = np.nan_to_num(power_curve_kw)
+        power_curve_kw = np.where(power_curve_kw<0,0,power_curve_kw)
+        power_curve_kw = np.where(power_curve_kw>turbine_specs["rated_power"],turbine_specs["rated_power"],power_curve_kw).tolist()
```

Comment:
These four lines could just exist in a code block for `if has_power_curve` since they're repeated in both blocks for `has_cp_curve` and `not has_cp_curve`; is that correct?

In general, I think the logic in these blocks could be cleaned up or simplified; I'd suggest taking another look at this flow to make sure it's accomplishing what you want in the most clear way

Reply:
Love this cleaned up version, thanks!